Invalidate prepared statements for views when their schema changes.

eliransin opened the pull request:
When a base table changes and altered, so does the views that might
refer to the added column (which includes "SELECT *" views and also
views that might need to use this column for rows lifetime (virtual
columns).
However the query processor implementation for views change notification
was an empty function.
Since views are tables, the query processor needs to at least treat them
as such (and maybe in the future, do also some MV specific stuff).
This commit adds a call to `on_update_column_family` from within
`on_update_view`.
The side effect true to this date is that prepared statements for views
which changed due to a base table change will be invalidated.

Fixes https://github.com/scylladb/scylladb/issues/16392

This series also adds a test which fails without this fix and passes when the fix is applied.

eliransin commented:
@nyh following your very good questions, here is a PR that fixes #16392 in what I believe to be the correct approach :slightly_smiling_face:

avikivity commented:
Do we allow materialized views defined as "SELECT *"?

eliransin commented:
> Do we allow materialized views defined as "SELECT *"?

Yes, we even have a member function to test for it, which is used in the alter table statement:
https://github.com/scylladb/scylladb/blob/b1ba904c49771abc0d8620af3bec1239b76bed14/cql3/statements/alter_table_statement.cc#L198

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5956/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5956/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7289/)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider8.cloudius-systems.com

avikivity commented on test/cql-pytest/test_materialized_view.py:
```diff
@@ -808,3 +808,14 @@ def test_mv_with_only_primary_key_rows(scylla_only, cql, test_keyspace):
             # We now believe that empty value serialization/deserialization is correct
+
+def test_mv_prepared_statement_with_altered_base(scylla_only, cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, 'id int PRIMARY KEY, v1 int') as base:
```
Does this test fail before the fix? If so it should be marked xfail (and the xfail removed in the next patch), so it can be bisected.

eliransin replied:
Sorry, I will revert the order and also remove the scylla_only flag as @nyh requested in the previous cycle.

eliransin commented:
Pushed new version:
1. Changed the order  between test and fix so branch will remain bisectable.
2. Added description for the test and remove the "scylla_only" flag.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5959/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7292/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5959/)
#### Build Details:
- Duration: 1 hr 51 min
- Builder: spider6.cloudius-systems.com

nyh commented on cql3/query_processor.cc:
```diff
@@ -1084,6 +1084,9 @@ void query_processor::migration_subscriber::on_update_aggregate(const sstring& k
         const sstring& view_name, bool columns_changed) {
+    // scylladb/scylladb#16392 - Materialized views are also tables so we need at least handle
+    // them as such when changed.
+    on_update_column_family(ks_name, view_name, columns_changed);
```
Very nice catch.

I want to point out to whoever is the expert in these "on_*" things that they are very fragile, poorly documented and very easy to misuse. About a month ago in 88a5ddabcec625434b2df0ed82a63c36245af35f I had to correct a very similar bug with creation of materialized views - these didn't get tablets created for them because the tablet creation was moved into a `on_before_create_column_family()` callback, and this wasn't called for new views, even though they are column families too.

In my opinion, these callbacks sound on paper like they are a good idea because it allows "mixing" different concerns easily - e.g., the code that deals with creating a table doesn't need to know about the existence of tablets. But in reality, the developers don't know about all these invisible assumptions and often make bugs - the bug that Eliran found here probably existed for at least 7 years (maybe we even copied this design from Cassandra?)

@tgrabiec you may be interested in this rant :-)

avikivity replied:
What's the alternative?

nyh replied:
The alternative is more "monolithic" code - you would have a "create table" code which also create the tablets, and since creating a view **also** creates a table (it doesn't duplicate the entire table creation code), it would naturally call the table-creation function and everything related to creating tables - including creating their tablets - would be done.
Anyway, I'm not saying I have the perfect solution. Just that it's the second bug we solved in a month that ended up being table-related callbacks which we forgot to call for the case of views (which, as Eliran noted, are also tables).

avikivity replied:
I think it's unrealistic to glue everything directly, and doesn't guarantee we don't forget to add the glue.
